runtime_lib/controller/shell.cpp

```diff
@@ -0,0 +1,167 @@
+static void command_help(char *line);
+static void command_x(char *line);
+
+static int new_cmd = 1;
+static int cmd_len;
+static char line[MAX_LINE_LENGTH];
+static command_entry command_tbl[] = {
+    {"cp", command_cp},
+    {"help", command_help},
+    {"x", command_x},
+};
+
+static char to_ascii(uint8_t byte) {
+  if ((byte >= 'A' && byte <= 'Z') || (byte >= 'a' && byte <= 'z') ||
+      (byte >= '0' && byte <= '9'))
+    return byte;
+
+  return '.';
+}
+
+/*
+        Print information about the command processors
+*/
+static void command_cp(char *line) {
+  uint32_t cp_count;
+  uint64_t base = get_base_address();
+
+  cp_count = *(uint32_t *)(base + 0x208);
+  xil_printf("Number of command processors: %u\r\n", cp_count);
+
+  uint64_t *queue_base = (uint64_t *)base;
+  for (uint32_t idx = 0; idx < cp_count; idx++) {
+    xil_printf("[%02u]\t0x%llx\r\n", idx, queue_base[idx]);
+  }
+}
+
+/*
+        List all supported shell commands
+*/
+static void command_help(char *line) {
+  uint32_t ll = 0; // line length
+  uint32_t len;    // string length
+
+  xil_printf("Supported commands:\r\n");
+
+  for (uint32_t idx = 0; idx < NUM_CMDS; idx++) {
+    len = strlen(command_tbl[idx].cmd) + 1;
+    if (ll + len > 80) {
+      ll = len;
+      xil_printf("\r\n");
+    }
+    xil_printf("%s ", command_tbl[idx].cmd);
+  }
+  xil_printf("\r\n");
+}
+
+/*
+        Print out a memory range
+
+        parameter 1: ARM virtual address (in hex or decimal)
+        parameter 2: number of bytes to read (should be aligned to 8)
+*/
+static void command_x(char *line) {
+  char *cmd;
+  char *address_str;
+  char *range_str;
+  uint64_t address = 0x400000;
+  uint32_t range = 8;
+
+  cmd = strtok(line, " ");
+  address_str = strtok(NULL, " ");
+  range_str = strtok(NULL, " ");
+
+  if (address_str)
+    address = strtoul(address_str, NULL, 16);
+  if (range_str)
+    range = strtoul(range_str, NULL, 16);
+
+  xil_printf("%s: addr=0x%llx range=0x%llx\r\n", cmd, address, range);
+
+  uint8_t *w = (uint8_t *)address;
+
+  for (uint32_t i = 0; i < range; i += 8)
+    xil_printf("[%16llx]: %02x %02x %02x %02x %02x %02x %02x %02x | "
+               "%c%c%c%c%c%c%c%c\r\n",
+               address + i * sizeof(*w), w[i], w[i + 1], w[i + 2], w[i + 3],
+               w[i + 4], w[i + 5], w[i + 6], w[i + 7], to_ascii(w[i]),
+               to_ascii(w[i + 1]), to_ascii(w[i + 2]), to_ascii(w[i + 3]),
+               to_ascii(w[i + 4]), to_ascii(w[i + 5]), to_ascii(w[i + 6]),
+               to_ascii(w[i + 7]));
+}
+
+static void handle_command(void) {
+  // look up the command and call the handler
+  for (uint32_t idx = 0; idx < NUM_CMDS; idx++) {
+    if (strncmp(line, command_tbl[idx].cmd, strlen(command_tbl[idx].cmd)) ==
+        0) {
+      command_tbl[idx].fn(line);
+      break;
+    }
+  }
+}
+
+void shell(void) {
+  uint8_t in;
+
+  if (new_cmd) {
+    new_cmd = 0;
+    cmd_len = 0;
+
+    // good old DOS prompt
+    xil_printf("C:\\> ");
```

Comment:
why?

Reply:
This is an interactive shell, so it needs a prompt to tell users that they can enter commands. This particular prompt is chosen to be distinguishable from bash or other commonly used shells. When a user has many windows open, it helps identify that this shell is the UART to the device (as opposed to the the host machine).

Reply:
I understand why there is a prompt. I don't understand why it's a fake DOS prompt instead of something that identifies what it actually is.

Reply:
I explained: this is to differentiate it from bash that uses % or $. If you want to pick a different character I'm open to suggestions.